iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs

```diff
@@ -322,65 +331,107 @@ public override async Task DisableTwinPatchAsync(CancellationToken cancellationT
             }
         }
 
-        public override async Task<Twin> SendTwinGetAsync(CancellationToken cancellationToken)
+        public override async Task<ClientTwin> GetTwinAsync(CancellationToken cancellationToken)
         {
             if (Logging.IsEnabled)
-                Logging.Enter(this, cancellationToken, nameof(SendTwinGetAsync));
+                Logging.Enter(this, cancellationToken, nameof(GetTwinAsync));
 
             try
             {
                 await EnableTwinPatchAsync(cancellationToken).ConfigureAwait(false);
 
-                Twin twin = await RoundTripTwinMessageAsync(AmqpTwinMessageType.Get, null, cancellationToken)
+                AmqpMessage responseFromService = await RoundTripTwinMessageAsync(AmqpTwinMessageType.Get, null, cancellationToken)
                     .ConfigureAwait(false);
-                return twin ?? throw new InvalidOperationException("Service rejected the message");
+
+                if (responseFromService == null)
+                {
+                    throw new InvalidOperationException("Service rejected the message");
+                }
+
+                using var reader = new StreamReader(responseFromService.BodyStream, _payloadConvention.PayloadEncoder.ContentEncoding);
+                string body = reader.ReadToEnd();
+
+                try
+                {
+                    ClientTwinProperties clientTwinProperties = _payloadConvention
+                        .PayloadSerializer
+                        .DeserializeToType<ClientTwinProperties>(body);
+
+                    var twinDesiredProperties = new DesiredPropertyCollection(clientTwinProperties.Desired)
+                    {
+                        PayloadConvention = _payloadConvention,
+                    };
+
+                    var twinReportedProperties = new ReportedPropertyCollection(clientTwinProperties.Reported)
+                    {
+                        PayloadConvention = _payloadConvention,
+                    };
+
+                    return new ClientTwin(twinDesiredProperties, twinReportedProperties);
+                }
+                catch (JsonReaderException ex)
```

Comment:
What if the user chooses a different library for de/serialization?

Reply:
we deserialize to an sdk-type that has newtonsoft json annotations. We do this because we know what keys to expect (desired, reported) as service defines them. The next level of deserialization will depend on the user's serializer.